apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/source/RdaVersion.java

```diff
@@ -0,0 +1,172 @@
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/** A utility class for comparing version numbers. */
+public class RdaVersion {
+
+  /** Name for the type group of the RegEx. */
+  private static final String TYPE_GROUP = "type";
+  /** Name for the major group of the RegEx. */
+  private static final String MAJOR_GROUP = "major";
+  /** Name for the minor group of the RegEx. */
+  private static final String MINOR_GROUP = "minor";
+  /** Name for the patch group of the RegEx. */
+  private static final String PATCH_GROUP = "patch";
+  /** A RegEx used for matching a versioning string. */
+  private static final Pattern VERSION_PATTERN =
+      Pattern.compile(
+          String.format(
+              "^(?<%s>[~^])?(?<%s>\\d+)\\.(?<%s>\\d+)\\.(?<%s>\\d+)$",
+              TYPE_GROUP, MAJOR_GROUP, MINOR_GROUP, PATCH_GROUP));
+
+  /** The major version (first number). */
+  private final int major;
+  /** The minor version (middle number). */
+  private final int minor;
+  /** The patch version (last number). */
+  private final int patch;
+
+  /**
+   * The required {@link CompatibilityLevel}, only used when comparing another {@link RdaVersion} to
+   * this one.
+   */
+  private final CompatibilityLevel compatability;
+
+  /**
+   * Builds an {@link RdaVersion} instance using the given version specifications.
+   *
+   * @param major The major version
+   * @param minor The minor version
+   * @param patch The patch version
+   * @param compatability The level requirement
+   */
+  private RdaVersion(int major, int minor, int patch, CompatibilityLevel compatability) {
+    this.major = major;
+    this.minor = minor;
+    this.patch = patch;
+    this.compatability = compatability;
+  }
+
+  /**
+   * Creates a {@link RdaVersionBuilder} to be used to build an {@link RdaVersion} object.
+   *
+   * @return A {@link RdaVersionBuilder}.
+   */
+  public static RdaVersionBuilder builder() {
+    return new RdaVersionBuilder();
+  }
+
+  /**
+   * Determines if this {@link RdaVersion} specifies a version or range that would accept the given
+   * {@link RdaVersion}.
+   *
+   * @param rdaVersionString The version being checked to see if it falls within range.
+   * @return True if the given {@link RdaVersion} is within range of this instance's {@link
+   *     RdaVersion} requirement.
+   */
+  public boolean allows(String rdaVersionString) {
+    // When we connect to S3 or some other source, there is a prefix on the version to distinguish
+    // it.
+    String[] bits = rdaVersionString.split(":");
+    String versionString = bits[bits.length - 1];
+    // Parse the version
+    RdaVersion rdaVersion = builder().versionString(versionString).build();
+
+    boolean compatable;
+
+    switch (compatability) {
+      case MAJOR:
+        compatable =
+            rdaVersion.major == major
+                && ((rdaVersion.minor == minor && rdaVersion.patch >= patch)
+                    || rdaVersion.minor > minor);
```

Comment:
Per semver rules, shouldnt a mismatch on minor and patch not cause breaking changes? Is this just being extra careful? Seems like if we're matching on major, we'd just check the major was good, and even if they are beyond on the minor/patch it should still work theoretically.

Reply:
We were following thn npm version range syntax:  https://semver.npmjs.com

Reply:
But this can be called on a non-range version too, per the docs? In which case even the above spec just uses normal semver (per https://semver.org/ , linked in that npm semver page). Seems like it'd need to discern if a range is being used, then use the more 'restrictive' check if so, else just use normal semver where a major would accept all minors and patches as compatible.

Just basing this off the test and hardcoded proto version not using a range ~/^ indicator.

Reply:
We chose to allow either an exact match (no `^` or `~`) or a range match.  The default config setting is using `~` so if no environment variable is set to override that we will be using a range.

Reply:
This logic, and the way we intend to use it, would be such that if we define `^0.12.1`, the code will continue to work all the way up to `1.0.0`.  It will fail if we get a `1.x.x` or if we get anything below `0.12.1` (even `0.12.0`), which is the desired behavior.

It's unlikely we'll use anything but the `^`, which matches only higher or equal versions within the same major, but we wanted to make it a little more flexible / easily extendable just in case.

Reply:
Ok I see, there's an implicit default exact match. Could we document the above information in one of the relevant locations?